### PR TITLE
TUI: detect background color

### PR DESCRIPTION
**Pull request by justinmk**
Merges https://github.com/neovim/neovim/pull/5027

- [x] fix `test/functional/terminal/tui_spec.lua` failures
- [x] lint
- [x] raise `OptionSet` if the response changes 'bg' (this matches Vim behavior)

Test case:

```
nvim -u NORC
:set bg?
```

- Works in iTerm2, but not Terminal.app
- Obviously it's racy because the terminal response is asynchronous, so `nvim -u NORC -c "set bg?"` will not reflect the response.
- ~~`OptionSet` is not raised because option.c checks `(!starting)`.~~

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -108,6 +108,7 @@ else()
 endif()
 if(CMAKE_BUILD_TYPE MATCHES Debug)
   set(DEBUG 1)
+  add_definitions(-DABORT_ON_INTERNAL_ERROR)
```
Comment (justinmk) [inline CMakeLists.txt]: @neovim/core FYI, setting `ABORT_ON_INTERNAL_ERROR` for "Debug" build-type by default.

**Comment by blueyed** [comment]
With https://github.com/neovim/neovim/commit/d3dc94615 Neovim hangs for me during startup, when opening an existing file, e.g. `nvim /etc/fstab`.

This is triggered by https://github.com/airblade/vim-gitgutter, using a minimal vimrc from its working tree:
```vim
let script_dir = fnamemodify(expand('<sfile>'), ':h')
let &rtp = script_dir.','.&rtp.','.script_dir.'/after'
```

(The stacktrace is from using my full config)
```
#0  0x00007fe798700c7e in epoll_pwait () from /usr/lib/libc.so.6
#1  0x00007fe79906737a in uv.io_poll () from /usr/lib/libuv.so.1
#2  0x00007fe799058ab0 in uv_run () from /usr/lib/libuv.so.1
#3  0x00005648524ecba2 in loop_poll_events (loop=0x5648527c5900 <main_loop>, ms=5)
    at ../src/nvim/event/loop.c:58
#4  0x000056485251acc5 in do_sleep (msec=5) at ../src/nvim/ex_docmd.c:7389
#5  0x000056485251ac23 in ex_sleep (eap=0x7ffde0521430) at ../src/nvim/ex_docmd.c:7378
#6  0x000056485250f698 in do_one_cmd (cmdlinep=0x7ffde0521630, flags=7, 
    cstack=0x7ffde0521740, fgetline=0x56485250c9b5 <get_loop_line>, cookie=0x7ffde05216e0)
    at ../src/nvim/ex_docmd.c:2242
#7  0x000056485250bd2b in do_cmdline (cmdline=0x0, fgetline=0x5648524d3169 <get_func_line>, 
    cookie=0x564853191b60, flags=7) at ../src/nvim/ex_docmd.c:593
#8  0x00005648524d2372 in call_user_func (fp=0x5648531cbca0, argcount=3, 
    argvars=0x7ffde05222e0, rettv=0x7ffde05229a0, firstline=2, lastline=2, selfdict=0x0)
    at ../src/nvim/eval.c:21574
#9  0x00005648524ad726 in call_func (funcname=0x56485316ffd0 "gitgutter#diff#run_diff", 
    len=23, rettv=0x7ffde05229a0, argcount_in=3, argvars_in=0x7ffde05222e0, argv_func=0x0, 
    firstline=2, lastline=2, doesrange=0x7ffde052249c, evaluate=true, partial=0x0, 
    selfdict_in=0x0) at ../src/nvim/eval.c:6383
#10 0x00005648524ace0f in get_func_tv (name=0x56485316ffd0 "gitgutter#diff#run_diff", 
    len=23, rettv=0x7ffde05229a0, arg=0x7ffde0522940, firstline=2, lastline=2, 
    doesrange=0x7ffde052249c, evaluate=1, partial=0x0, selfdict=0x0)
    at ../src/nvim/eval.c:6144
#11 0x00005648524a82f3 in eval7 (arg=0x7ffde0522940, rettv=0x7ffde05229a0, evaluate=1, 
    want_string=0) at ../src/nvim/eval.c:4245
#12 0x00005648524a79f5 in eval6 (arg=0x7ffde0522940, rettv=0x7ffde05229a0, evaluate=1, 
    want_string=0) at ../src/nvim/eval.c:3942
#13 0x00005648524a74e1 in eval5 (arg=0x7ffde0522940, rettv=0x7ffde05229a0, evaluate=1)
    at ../src/nvim/eval.c:3793
#14 0x00005648524a674c in eval4 (arg=0x7ffde0522940, rettv=0x7ffde05229a0, evaluate=1)
    at ../src/nvim/eval.c:3530
#15 0x00005648524a6574 in eval3 (arg=0x7ffde0522940, rettv=0x7ffde05229a0, evaluate=1)
    at ../src/nvim/eval.c:3448
#16 0x00005648524a63cb in eval2 (arg=0x7ffde0522940, rettv=0x7ffde05229a0, evaluate=1)
    at ../src/nvim/eval.c:3379
#17 0x00005648524a61e2 in eval1 (arg=0x7ffde0522940, rettv=0x7ffde05229a0, evaluate=1)
    at ../src/nvim/eval.c:3306
#18 0x00005648524a6111 in eval0 (
    arg=0x56485310b863 "gitgutter#diff#run_diff(a:bufnr, 'index', 0)", 
    rettv=0x7ffde05229a0, nextcmd=0x7ffde0522aa8, evaluate=1) at ../src/nvim/eval.c:3266
#19 0x00005648524a1a28 in ex_let (eap=0x7ffde0522aa0) at ../src/nvim/eval.c:1489
#20 0x000056485250f698 in do_one_cmd (cmdlinep=0x7ffde0522ca0, flags=7, 
    cstack=0x7ffde0522db0, fgetline=0x5648524d3169 <get_func_line>, cookie=0x56485319e960)
    at ../src/nvim/ex_docmd.c:2242
#21 0x000056485250bd2b in do_cmdline (cmdline=0x0, fgetline=0x5648524d3169 <get_func_line>, 
    cookie=0x56485319e960, flags=7) at ../src/nvim/ex_docmd.c:593
#22 0x00005648524d2372 in call_user_func (fp=0x56485319d640, argcount=2, 
    argvars=0x7ffde0523950, rettv=0x7ffde0523b30, firstline=2, lastline=2, selfdict=0x0)
    at ../src/nvim/eval.c:21574
#23 0x00005648524ad726 in call_func (funcname=0x56485319fd90 "gitgutter#process_buffer", 
    24, rettv=0x7ffde0523b30, argcount_in=2, argvars_in=0x7ffde0523950, argv_func=0x0, firstline=2, lastline=2, doesrange=0x7ffde0523afc, evaluate=true, partial=0x0, selfdict_in=0x0)
    at ../src/nvim/eval.c:6383
#24 0x00005648524ace0f in get_func_tv (name=0x56485319fd90 "gitgutter#process_buffer", len=24, rettv=0x7ffde0523b30, arg=0x7ffde0523b00, firstline=2, lastline=2, doesrange=0x7ffde0523afc, 
    evaluate=1, partial=0x0, selfdict=0x0) at ../src/nvim/eval.c:6144
#25 0x00005648524a4d7f in ex_call (eap=0x7ffde0523c60) at ../src/nvim/eval.c:2742
#26 0x000056485250f698 in do_one_cmd (cmdlinep=0x7ffde0523e60, flags=7, cstack=0x7ffde0523f70, fgetline=0x5648524d3169 <get_func_line>, cookie=0x56485315aab0) at ../src/nvim/ex_docmd.c:2242
#27 0x000056485250bd2b in do_cmdline (cmdline=0x0, fgetline=0x5648524d3169 <get_func_line>, cookie=0x56485315aab0, flags=7) at ../src/nvim/ex_docmd.c:593
#28 0x00005648524d2372 in call_user_func (fp=0x564852f135d0, argcount=0, argvars=0x7ffde0524b10, rettv=0x7ffde0524cf0, firstline=2, lastline=2, selfdict=0x0) at ../src/nvim/eval.c:21574
#29 0x00005648524ad726 in call_func (funcname=0x564853199320 "\200\375R92_on_bufenter", len=17, rettv=0x7ffde0524cf0, argcount_in=0, argvars_in=0x7ffde0524b10, argv_func=0x0, firstline=2, 
    lastline=2, doesrange=0x7ffde0524cbc, evaluate=true, partial=0x0, selfdict_in=0x0) at ../src/nvim/eval.c:6383
#30 0x00005648524ace0f in get_func_tv (name=0x564853199320 "\200\375R92_on_bufenter", len=17, rettv=0x7ffde0524cf0, arg=0x7ffde0524cc0, firstline=2, lastline=2, doesrange=0x7ffde0524cbc, 
    evaluate=1, partial=0x0, selfdict=0x0) at ../src/nvim/eval.c:6144
#31 0x00005648524a4d7f in ex_call (eap=0x7ffde0524e20) at ../src/nvim/eval.c:2742
#32 0x000056485250f698 in do_one_cmd (cmdlinep=0x7ffde0525020, flags=7, cstack=0x7ffde0525130, fgetline=0x564852549b2d <getnextac>, cookie=0x7ffde0525720) at ../src/nvim/ex_docmd.c:2242
#33 0x000056485250bd2b in do_cmdline (cmdline=0x0, fgetline=0x564852549b2d <getnextac>, cookie=0x7ffde0525720, flags=7) at ../src/nvim/ex_docmd.c:593
#34 0x000056485254955b in apply_autocmds_group (event=EVENT_BUFENTER, fname=0x5648531a1c60 "/etc/fstab", fname_io=0x0, force=false, group=-3, buf=0x564852a2cc20, eap=0x0)
    at ../src/nvim/fileio.c:7023
#35 0x0000564852548c2c in apply_autocmds (event=EVENT_BUFENTER, fname=0x0, fname_io=0x0, force=false, buf=0x564852a2cc20) at ../src/nvim/fileio.c:6655
#36 0x000056485257bcfe in main (argc=2, argv=0x7ffde0525a88) at ../src/nvim/main.c:515
```

`Ctrl-c` can be used to continue, but causes e.g. `nvim -d 1 2` to compare 1 with itself only.

vim-gitgutter appears to loop here:
https://github.com/airblade/vim-gitgutter/blob/5c8efd3625b7f5a586ad3ce5922ad45ae5661393/autoload/gitgutter/diff.vim#L72
via `function <SNR>92_on_bufenter[6]..gitgutter#process_buffer[8]..gitgutter#diff#run_diff`

The problem is that it triggers a job, which would set the expected var, but this likely is not being run by Neovim then with/after this patch.

https://github.com/airblade/vim-gitgutter/blob/5c8efd3625b7f5a586ad3ce5922ad45ae5661393/autoload/gitgutter/utility.vim#L126-L145

**Comment by blueyed** [comment]
Oh, missed https://github.com/neovim/neovim/issues/9675.

**Comment by will** [comment]
This is an awesome feature, thank you!

**Pull request merged**